slackapi/bolt-js
# Add in check-update hook functionality for Bolt JS

###  Summary

Adding in `check-update` hook functionality for Bolt JS projects on the CLI. 

**TODOs**

- [x] ~Check to see what our backup strategy should be if folks don't have a `package-lock.json` in their project~ In favor of code simplicity, added a `package.json` only approach since it's pretty much for sure going to be there, but can add `package-lock.json` back in if we feel like it's needed
- [x] Think through some error handling scenarios, if applicable
- [x] Look into build error - testing error showing up only on Github
- [x] Linting for test and `check-update` hook file
- [x] ~Look into using `rewire` for mocking function / module responses rather than `sinon`~ Was having a hard time getting this working, sadly 😢 ended up keeping the `sinon` stubbing 
- [x] Simplify any code possible in hook + testing files


**Testing instructions** 
1. Pull down this branch (`next-gen-check-update`) and run `npm link` 
2. Once that finishes, create a Bolt JS test project using an existing template (I used `slack-samples/bolt-js-request-time-off`)
3. Run `npm link @slack/bolt` in the Bolt JS test project directory to link the local branch with the Bolt package used in the test project
4. Run `slack doctor` to test - you should see a "Dependencies" section in the output under "PROJECT" that lists out `@slack/bolt`
<img width="396" alt="Screen Shot 2022-10-06 at 9 54 24 AM" src="https://user-images.githubusercontent.com/12901850/194331642-a7580b29-b7f4-404d-87dd-2b2ba44fc8ac.png">
5. To test the logic of the hook, update the `package.json` version of `@slack/bolt`. **Please replace the `next-gen` dist tag in your project's `package.json` and swap it out so that it uses an exact version of the package (ex: `4.0.0-nextGen.6`) - this is for local testing only because it is unable to pull the dist tag version from the symlinked repo. You will also need to update the `package.json` in your local `bolt-js` project so that the package version matches whatever you are testing with.** This is because a symlink is being used and not having consistent versioning across the `bolt-js` package and the project's `package.json` will throw an error and cause no terminal output.

Example output here if an update is available for `slack doctor`:
<img width="868" alt="Screen Shot 2022-10-05 at 3 55 32 PM" src="https://user-images.githubusercontent.com/12901850/194347517-c3921758-36d8-4252-89c2-ff17027130fa.png">
<img width="705" alt="Screen Shot 2022-10-06 at 5 51 54 PM" src="https://user-images.githubusercontent.com/12901850/194429464-d1dc533c-7805-45dd-a94c-d2d8bac9e5a0.png">

Example output if updates are available for `slack update`:
<img width="636" alt="Screen Shot 2022-10-11 at 3 36 13 PM" src="https://user-images.githubusercontent.com/12901850/195183017-e4a17334-e57b-4f9d-bed0-95288c1bba3c.png">

### Requirements (place an `x` in each `[ ]`)
